porch/controllers/packagevariantsets/pkg/controllers/packagevariantset/render.go

```diff
@@ -0,0 +1,315 @@
+	// on a per-GVK basis
+	result["name"] = u.GetName()
+	result["namespace"] = u.GetNamespace()
+	result["labels"] = u.GetLabels()
+	result["annotations"] = u.GetAnnotations()
+
+	return result, nil
+}
+
+func copyAndOverlayMapExpr(fieldName string, inMap map[string]string, mapExprs []api.MapExpr, inputs map[string]interface{}) (map[string]string, error) {
+	outMap := make(map[string]string, len(inMap))
+	for k, v := range inMap {
+		outMap[k] = v
+	}
+
+	var err error
+	for i, me := range mapExprs {
+		var k, v string
+		if me.Key != nil {
+			k = *me.Key
+		}
+		if me.KeyExpr != nil {
+			k, err = evalExpr(*me.KeyExpr, inputs)
+			if err != nil {
+				return nil, fmt.Errorf("%s[%d].keyExpr: %s", fieldName, i, err.Error())
+			}
+		}
+		if me.Value != nil {
+			v = *me.Value
+		}
+		if me.ValueExpr != nil {
+			v, err = evalExpr(*me.ValueExpr, inputs)
+			if err != nil {
+				return nil, fmt.Errorf("%s[%d].valueExpr: %s", fieldName, i, err.Error())
+			}
+		}
+		outMap[k] = v
+	}
+
+	if len(outMap) == 0 {
+		return nil, nil
+	}
+
+	return outMap, nil
+}
+
+func copyAndOverlayStringSlice(fieldName string, in, exprs []string, inputs map[string]interface{}) ([]string, error) {
+	outMap := make(map[string]bool, len(in)+len(exprs))
+
+	for _, v := range in {
+		outMap[v] = true
+	}
+	for i, e := range exprs {
+		v, err := evalExpr(e, inputs)
+		if err != nil {
+			return nil, fmt.Errorf("%s[%d]: %s", fieldName, i, err.Error())
+		}
+		outMap[v] = true
+	}
+
+	if len(outMap) == 0 {
+		return nil, nil
+	}
+
+	var out []string
+	for k := range outMap {
+		out = append(out, k)
+	}
+	sort.Strings(out)
+	return out, nil
+}
+
+func evalExpr(expr string, inputs map[string]interface{}) (string, error) {
+	prog, err := compileExpr(expr)
+	if err != nil {
+		return "", err
+	}
+
+	val, _, err := prog.Eval(inputs)
+	if err != nil {
+		return "", err
+	}
+
+	result, err := val.ConvertToNative(reflect.TypeOf(""))
+	if err != nil {
+		return "", err
+	}
+
+	s, ok := result.(string)
+	if !ok {
+		return "", fmt.Errorf("expression returned non-string value: %v", result)
+	}
+
+	return s, nil
+}
+
+// compileExpr returns a compiled CEL expression.
+func compileExpr(expr string) (cel.Program, error) {
+	var opts []cel.EnvOption
+	opts = append(opts, cel.HomogeneousAggregateLiterals())
+	opts = append(opts, cel.EagerlyValidateDeclarations(true), cel.DefaultUTCTimeZone(true))
+	// TODO: uncomment after updating to latest k8s
+	//opts = append(opts, library.ExtensionLibs...)
+	opts = append(opts, cel.Variable(RepoDefaultVarName, cel.StringType))
+	opts = append(opts, cel.Variable(PackageDefaultVarName, cel.StringType))
+	opts = append(opts, cel.Variable(UpstreamVarName, cel.DynType))
+	opts = append(opts, cel.Variable(TargetVarName, cel.DynType))
+	opts = append(opts, cel.Variable(RepositoryVarName, cel.DynType))
+
+	env, err := cel.NewEnv(opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	ast, issues := env.Compile(expr)
```

Comment:
Looks great. Once https://github.com/kubernetes/kubernetes/pull/116779 merges, you'll also have the option to more easily use a base environment provided by k8s (which as a few more libs and will evolve over time).  You'd also get some versioning support, if that is useful to you.

Reply:
Cool, thanks for taking a look, Joe.